activation/validation.go

```diff
@@ -62,31 +61,52 @@ func (v *Validator) NIPost(nodeId types.NodeID, commitmentAtxId types.ATXID, nip
 		return 0, err
 	}
 
+	if err := v.Post(nodeId, commitmentAtxId, nipost.Post, nipost.PostMetadata, numUnits, opts...); err != nil {
+		return 0, fmt.Errorf("invalid Post: %v", err)
+	}
+
 	var ref types.PoetProofRef
 	copy(ref[:], nipost.PostMetadata.Challenge)
-	proof, err := v.poetDb.GetProof(ref)
+	proof, statement, err := v.poetDb.GetProof(ref)
 	if err != nil {
 		return 0, fmt.Errorf("poet proof is not available %x: %w", nipost.PostMetadata.Challenge, err)
 	}
 
-	if !contains(proof, nipost.Challenge.Bytes()) {
-		return 0, fmt.Errorf("challenge is not included in the proof %x", nipost.PostMetadata.Challenge)
-	}
-
-	if err := v.Post(nodeId, commitmentAtxId, nipost.Post, nipost.PostMetadata, numUnits, opts...); err != nil {
-		return 0, fmt.Errorf("invalid Post: %v", err)
+	if err := validateMerkleProof(expectedChallenge[:], &nipost.Membership, statement[:]); err != nil {
+		return 0, fmt.Errorf("invalid membership proof %w", err)
 	}
 
 	return proof.LeafCount, nil
 }
 
-func contains(proof *types.PoetProof, member []byte) bool {
-	for _, part := range proof.Members {
-		if bytes.Equal(part[:], member) {
-			return true
+func validateMerkleProof(leaf []byte, proof *types.MerkleProof, expectedRoot []byte) error {
+	nodes := make([][]byte, 0, len(proof.Nodes))
+	for _, n := range proof.Nodes {
+		nodes = append(nodes, n.Bytes())
+	}
+	ok, err := merkle.ValidatePartialTree(
+		[]uint64{proof.LeafIndex},
+		[][]byte{leaf},
+		nodes,
+		expectedRoot,
+		poetShared.HashMembershipTreeNode,
+	)
+	if err != nil {
+		return fmt.Errorf("validating merkle proof: %w", err)
+	}
+	if !ok {
+		hexNodes := make([]string, 0, len(proof.Nodes))
+		for _, n := range proof.Nodes {
+			hexNodes = append(hexNodes, n.Hex())
 		}
+		return fmt.Errorf(
+			"invalid merkle proof, calculated root does not match the proof root, leaf: %v, nodes: %v, expected root: %v",
+			util.Encode(leaf),
+			hexNodes,
+			util.Encode(expectedRoot),
```

Comment:
what do you think about pattern in types/missing.go ? it allows to print error nicely by implementing logging interface

Reply:
I misunderstood what you meant. Do you mean to create an error type to represent a particular failure and implement logging i/f for it? It seems a bit of an overkill if the error is used only in one place and if it is not desired to handle it in any way (like in this case).